ably/rest_channel.go

```diff
@@ -46,17 +49,46 @@ func newRestChannel(name string, client *REST) *RestChannel {
 	return c
 }
 
-func (c *RestChannel) Publish(name string, data interface{}) error {
-	messages := []*proto.Message{
+// Publish publishes a message on the channel.
+func (c *RESTChannel) Publish(ctx context.Context, name string, data interface{}) error {
+	return c.PublishBatch(ctx, []*Message{
 		{Name: name, Data: data},
-	}
-	return c.PublishAll(messages)
+	})
+}
+
+// Message is what Ably channels send and receive.
+type Message = proto.Message
+
+// PublishBatch publishes multiple messages in a batch.
+func (c *RESTChannel) PublishBatch(ctx context.Context, messages []*Message) error {
+	return c.PublishBatchWithOptions(ctx, messages, RESTChannelPublishBatchOptions{})
+}
+
+// RESTChannelPublishBatchOptions are optional parameters for RESTChannel.PublishBatchWithOptions.
+type RESTChannelPublishBatchOptions []func(*restChannelPublishBatchOptions)
+
+type restChannelPublishBatchOptions struct {
+	params map[string]string
 }
 
-// PublishAll sends multiple messages in the same http call.
-// This is the more efficient way of transmitting a batch of messages
-// using the Rest API.
-func (c *RestChannel) PublishAll(messages []*proto.Message) error {
+// Params adds query parameters to the resulting HTTP request to the REST API.
+func (o RESTChannelPublishBatchOptions) Params(params map[string]string) RESTChannelPublishBatchOptions {
+	return append(o, func(options *restChannelPublishBatchOptions) {
+		options.params = params
+	})
+}
+
+// PublishBatchWithOptions is PublishBatch with optional parameters.
+func (c *RESTChannel) PublishBatchWithOptions(ctx context.Context, messages []*Message, options RESTChannelPublishBatchOptions) error {
```

Comment:
Why not use a variadic list of functional options rather that a struct?

With the current code, if I want to publish a batch with some params, I'd write:

```go
channel.PublishBatchWithOptions(
	ctx,
	msgs,
	ably.RESTChannelPublishBatchOptions{}.Params(map[string]string{"foo": "bar"}),
)
```

whereas with a variadic list it would be:

```go
channel.PublishBatchWithOptions(
	ctx,
	msgs,
	ably.PublishBatchParams(map[string]string{"foo": "bar"}),
)
```

(I also think there should be a singular form):

```go
channel.PublishBatchWithOptions(
	ctx,
	msgs,
	ably.PublishBatchParam("foo", "bar"),
)
```

Do you not find this to be a cleaner API? It's also inline with how this is often implemented in other popular API client libraries (e.g. [oauth2](https://godoc.org/golang.org/x/oauth2#Config.Exchange), [grpc](https://godoc.org/google.golang.org/grpc#ClientConn.Invoke) and [AWS SDK](https://docs.aws.amazon.com/sdk-for-go/api/service/sts/#STS.GetCallerIdentityWithContext)).

Reply:
ClientOptions is already done like this, and I just followed the pattern here. ([We discussed this back then.](https://github.com/ably/ably-go/pull/141#discussion_r401739264))

I still think that's the right option for ClientOptions, given there's tons of them. But for optional parameters, of which there shouldn't be too many, maybe it's worth polluting the package-level namespace to avoid the `ably.FooOptions{}.` noise.

I agree that the `RESTChannelPublishBatch` prefix is probably too much and just `PublishBatch` would be preferahyble.

Reply:
> ClientOptions is already done like this

My apologies, it looks like I made a similar comment [here](https://github.com/ably/ably-go/pull/145#discussion_r405653434) but failed to follow up, and I also made a similar comment more recently [here](https://github.com/ably/ably-go/issues/197#issuecomment-709661794), but I'll now make a more explicit comment in #197 so we can discuss this further and gather more feedback.

Reply:
I've continued this discussion here: https://github.com/ably/ably-go/issues/197#issuecomment-710844145

Reply:
Done at 7f31d1c.